src/v/iceberg/tests/values_json_test.cc

```diff
@@ -0,0 +1,911 @@
+  },
+  values_json_test_case{
+    "Expected fractional part for timestamptz_value, got '1971-01-01T00:00:00'",
+    timestamptz_type{},
+    R"("1971-01-01T00:00:00")",
+  },
+  values_json_test_case{
+    "Expected 6-digit microsecond resolution, got '023'",
+    timestamptz_type{},
+    R"("1971-01-01T00:00:00.023+00:00")",
+  },
+  values_json_test_case{
+    "Failed to parse microseconds: 'foobar'",
+    timestamptz_type{},
+    R"("1971-01-01T00:00:00.foobar+00:00")",
+  },
+  values_json_test_case{
+    "Expected offset part for timestamptz_value, got "
+    "'1971-01-01T00:00:00.000023'",
+    timestamptz_type{},
+    R"("1971-01-01T00:00:00.000023")",
+  },
+  values_json_test_case{
+    "Failed to parse date string 'junk1971-01-01T00:00:00', expected format "
+    "'%FT%T'",
+    timestamptz_type{},
+    R"("junk1971-01-01T00:00:00.000023+00:00")",
+  },
+  values_json_test_case{
+    "Expected 8 hex digits for fixed[4]: got 4",
+    fixed_type{4},
+    R"("ffff")",
+  },
+  values_json_test_case{
+    "Failed to parse uuid: invalid uuid string",
+    uuid_type{},
+    R"("ffff")",
+  },
+  values_json_test_case{
+    "Expected even length hex string, got 9",
+    binary_type{},
+    R"("123456789")",
+  },
+  values_json_test_case{
+    "Failed to parse hex byte 'GG' - ec: 'generic:22'",
+    binary_type{},
+    R"("GGGGGGGG")",
+  },
+  values_json_test_case{
+    fmt::format(
+      "Expected at most 5-byte precision for {}, got 6", decimal_type{5, 2}),
+    decimal_type{5, 2},
+    R"("1000.00")",
+  },
+  values_json_test_case{
+    "Failed to parse int128",
+    decimal_type{5, 2},
+    R"("FuB.Ar")",
+  },
+};
+
+} // namespace
+
+struct ValuesJsonParseErrorsTest : public ValuesJsonTestBase {};
+
+INSTANTIATE_TEST_SUITE_P(
+  ValuesJsonSerde, ValuesJsonParseErrorsTest, ::testing::ValuesIn(parse_errs));
+
+TEST_P(ValuesJsonParseErrorsTest, ParsingFailsGracefully) {
+    json::Document parsed;
+    parsed.Parse(get_json_rep().data());
+    ASSERT_THROW(
+      value_from_json(parsed, get_type(), field_required::yes),
+      std::invalid_argument);
+    ASSERT_THROW(
+      {
+          try {
+              value_from_json(parsed, get_type(), field_required::yes);
+          } catch (const std::invalid_argument& e) {
+              EXPECT_THAT(e.what(), testing::HasSubstr(get_description()));
+              throw;
+          }
+      },
+      std::invalid_argument);
+}
+
+namespace {
+struct decimal_parsing_test_case {
+    std::string description;
+    std::string json;
+    decimal_value value;
+    decimal_type type;
+    std::optional<std::string> re_json = std::nullopt;
+};
+
+std::ostream&
+operator<<(std::ostream& os, const decimal_parsing_test_case& tc) {
+    return os << tc.description;
+}
+
+struct DecimalRoundTripTest
+  : ::testing::Test
+  , testing::WithParamInterface<decimal_parsing_test_case> {};
+
+static const std::vector<decimal_parsing_test_case> decimal_cases{
+  decimal_parsing_test_case{
+    "simple",
```

Comment:
Not blocking, but given how non-trivial decimal parsing is, it would be great to test variants of 0 (0.0, 0., 0000., .00000, -0.0, etc). Wouldn't be surprised if that's a common default

Reply:
ya, great point. i'll bolt that onto the followup ticket

Reply:
which is here fyi: https://redpandadata.atlassian.net/browse/CORE-8835